datafusion/physical-expr/src/encoding_expressions.rs

```diff
@@ -0,0 +1,340 @@
+                    "Failed to decode value using hex: {}",
+                    e
+                ))
+            })?,
+        };
+
+        Ok(ColumnarValue::Scalar(ScalarValue::LargeBinary(Some(out))))
+    }
+
+    fn decode_binary_array<T>(self, value: &dyn Array) -> Result<ColumnarValue>
+    where
+        T: OffsetSizeTrait,
+    {
+        let input_value = as_generic_binary_array::<T>(value)?;
+        let array: ArrayRef = match self {
+            Self::Base64 => decode_to_array!(base64_decode, input_value),
+            Self::Hex => decode_to_array!(hex_decode, input_value),
+        };
+        Ok(ColumnarValue::Array(array))
+    }
+
+    fn decode_utf8_array<T>(self, value: &dyn Array) -> Result<ColumnarValue>
+    where
+        T: OffsetSizeTrait,
+    {
+        let input_value = as_generic_string_array::<T>(value)?;
+        let array: ArrayRef = match self {
+            Self::Base64 => decode_to_array!(base64_decode, input_value),
+            Self::Hex => decode_to_array!(hex_decode, input_value),
+        };
+        Ok(ColumnarValue::Array(array))
+    }
+}
+
+impl fmt::Display for Encoding {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", format!("{self:?}").to_lowercase())
+    }
+}
+
+impl FromStr for Encoding {
+    type Err = DataFusionError;
+    fn from_str(name: &str) -> Result<Encoding> {
+        Ok(match name {
+            "base64" => Self::Base64,
+            "hex" => Self::Hex,
+            _ => {
+                let options = [Self::Base64, Self::Hex]
+                    .iter()
+                    .map(|i| i.to_string())
+                    .collect::<Vec<_>>()
+                    .join(", ");
+                return Err(DataFusionError::Plan(format!(
+                    "There is no built-in encoding named '{name}', currently supported encodings are: {options}",
+                )));
+            }
+        })
+    }
+}
+
+/// Encodes the given data, accepts Binary, LargeBinary, Utf8 or LargeUtf8 and returns a [`ColumnarValue`].
+/// Second argument is the encoding to use.
+/// Standard encodings are base64 and hex.
+pub fn encode(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+    if args.len() != 2 {
+        return Err(DataFusionError::Internal(format!(
+            "{:?} args were supplied but encode takes exactly two arguments",
+            args.len(),
+        )));
+    }
+    let encoding = match &args[1] {
+        ColumnarValue::Scalar(scalar) => match scalar {
+            ScalarValue::Utf8(Some(method)) | ScalarValue::LargeUtf8(Some(method)) => {
+                method.parse::<Encoding>()
+            }
+            other => Err(DataFusionError::Internal(format!(
+                "Unsupported data type {other:?} for function encode",
+            ))),
+        },
+        ColumnarValue::Array(_) => Err(DataFusionError::Internal(
+            "Encode using dynamically decided method is not yet supported".into(),
+        )),
```

Comment:
Given a user can cause this error I don't think it is internal (aka it isn't a bug in datafusion, which is what internal errors are for).

Also I think it would help to make the error a bit more user friendly

```suggestion
        ColumnarValue::Array(_) => Err(DataFusionError::NotYetImplemented(
            "Second argument to encode must be a constant: Encode using dynamically decided method is not yet supported".into(),
                    )),
```